src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringSpace.java

```diff
@@ -0,0 +1,209 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.knn.plugin.script;
+
+import com.amazon.opendistroforelasticsearch.knn.index.KNNVectorFieldMapper;
+import com.amazon.opendistroforelasticsearch.knn.plugin.stats.KNNCounter;
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.index.mapper.BinaryFieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.NumberFieldMapper;
+import org.elasticsearch.script.ScoreScript;
+import org.elasticsearch.search.lookup.SearchLookup;
+
+import java.io.IOException;
+import java.util.Base64;
+import java.util.BitSet;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import static org.elasticsearch.index.mapper.NumberFieldMapper.NumberType.LONG;
+
+/**
+ * KNNScoringSpace is used to validate/prepare a user provided query and similarity space for knn scripting execution
+ * and provide the correct KNNScoringScript.
+ */
+public abstract class KNNScoringSpace {
+
+    protected Object processedQuery;
```

Comment:
why not make processedQuery as method and do validation inside it? it is kind of dangerous to call protected method inside constructor because you are calling an instance method before creating an instance.
IMO, KNNScoringSpace can be made interface and make helper method in a different class.

Reply:
Generally subclass are used when it naturally depicts parent -> child relationship where there are logics which are common to be inherited and some to be overridden.  Here, there is no strategy/algorithm that is common while child class can behave differently. I only see some helper methods that will be used by child class and the common method(getScoreScript) which will be called by outside is abstract.  Also, any reason why made L2/Hamming as inner class ? Interface is preferred if intention is that L2ScoringSpace/BitHammingScoringSpace are all ScoringSpace.
Did i convince you?